### ISSUE TITLE
False positive in `ptr_arg` when function pushes to `&mut Vec`

### ISSUE THREAD

**Issue by jorgecarleitao** [rust-lang/rust-clippy#8463]
### Summary

When a function has the signature `&mut Vec<T>` and pushes to it, it cannot be converted to `&mut [T]`. However, in some cases, clippy recommends doing so.

### Lint Name

ptr_arg

### Reproducer

It is not easy to trigger this, so I will use the example I found it:

```
git clone git@github.com:jorgecarleitao/arrow2.git
cargo clippy --features full
```

I saw this happen:

```
warning: writing `&mut Vec` instead of `&mut [_]` involves a new object where a slice will do
   --> src/io/ipc/write/serialize.rs:209:12
    |
209 |     nodes: &mut Vec<ipc::FieldNode>,
    |            ^^^^^^^^^^^^^^^^^^^^^^^^ help: change this to: `&mut [ipc::FieldNode]`
    |
    = note: `#[warn(clippy::ptr_arg)]` on by default
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#ptr_arg

warning: writing `&mut Vec` instead of `&mut [_]` involves a new object where a slice will do
   --> src/io/ipc/write/serialize.rs:298:12
    |
298 |     nodes: &mut Vec<ipc::FieldNode>,
    |            ^^^^^^^^^^^^^^^^^^^^^^^^ help: change this to: `&mut [ipc::FieldNode]`
    |
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#ptr_arg

warning: writing `&mut Vec` instead of `&mut [_]` involves a new object where a slice will do
   --> src/io/ipc/write/serialize.rs:341:12
    |
341 |     nodes: &mut Vec<ipc::FieldNode>,
    |            ^^^^^^^^^^^^^^^^^^^^^^^^ help: change this to: `&mut [ipc::FieldNode]`
    |
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#ptr_arg

warning: writing `&mut Vec` instead of `&mut [_]` involves a new object where a slice will do
   --> src/io/ipc/write/serialize.rs:642:14
    |
642 |     buffers: &mut Vec<ipc::Buffer>,
    |              ^^^^^^^^^^^^^^^^^^^^^ help: change this to: `&mut [ipc::Buffer]`
    |
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#ptr_arg

warning: `arrow2` (lib) generated 4 warnings
```

However, If any of them are changed as suggested, the code does not compile.

The image below quickly illustrates the issue: `buffer.push` is not implemented for `&mut [ipc::Buffer]`, but clippy recommends replacing it:

<img width="795" alt="Screenshot 2022-02-23 at 05 12 51" src="https://user-images.githubusercontent.com/2772607/155260677-2a180f3d-6045-4a77-8032-ff7b9a0eee66.png">

### Version

```text
rustc 1.61.0-nightly (45e2c2881 2022-02-20)
binary: rustc
commit-hash: 45e2c2881d11324d610815bfff097e25c412199e
commit-date: 2022-02-20
host: x86_64-unknown-linux-gnu
release: 1.61.0-nightly
LLVM version: 14.0.0
```

**Comment by jorgecarleitao** [comment]
@Jarcho, has this been fixed and released? This is now showing in rust stable 1.60.

**Comment by Jarcho** [comment]
I think it just missed the beta branch in February, so it should be in `1.61`.

**Comment by eirnym** [comment]
Still is an issue in Rust `1.96.0`
